File: compatibility/pg/node/pg_test.js
const { Client } = require('pg');
const fs = require('fs');

class Test {
    constructor(query, expectedResults) {
        this.query = query;
        this.expectedResults = expectedResults;
    }

    async run(client) {
        try {
            console.log(`Running test: ${this.query}`);
            const res = await client.query(this.query);
            if (res.rows.length === 0) {
                console.log("Returns 0 rows");
                return this.expectedResults.length === 0;
            }
            if (res.fields.length !== this.expectedResults[0].length) {
                console.error(`Expected ${this.expectedResults[0].length} columns, got ${res.fields.length}`);
                return false;
            }
            for (let i = 0; i < res.rows.length; i++) {
                const row = res.rows[i];
                const expectedRow = this.expectedResults[i];
                for (let j = 0; j < expectedRow.length; j++) {
                    const columnName = res.fields[j].name;
                    if (String(row[columnName]) !== String(expectedRow[j])) {
                        console.error(`Expected: ${String(expectedRow[j])}, got: ${String(row[columnName])}`);
                        return false;
                    }
                }
            }
            console.log(`Returns ${res.rows.length} rows`);
            if (res.rows.length !== this.expectedResults.length) {
                console.error(`Expected ${this.expectedResults.length} rows`);
                return false;
            }
            return true;
        } catch (err) {
            console.error(err.message);
            return false;
        }
    }
}

class Tests {
    constructor() {
        this.client = null;
        this.tests = [];
    }

    async connect(ip, port, user, password) {
        this.client = new Client({
            host: ip,
            port: port,
            user: user,
            password: password,
            database: 'postgres'
        });
        await this.client.connect();
    }

    async disconnect() {
        await this.client.end();
    }

    addTest(query, expectedResults) {
        this.tests.push(new Test(query, expectedResults));
    }

    async runTests() {
        for (const test of this.tests) {
            if (!(await test.run(this.client))) {
                return false;
            }
        }
        return true;
    }

    readTestsFromFile(filename) {
        const data = fs.readFileSync(filename, 'utf8');
        const lines = data.split('\n');
        let query = null;
        let expectedResults = [];
        for (const line of lines) {
            const trimmedLine = line.trim();
            if (!trimmedLine) {
                if (query) {
                    this.addTest(query, expectedResults);
                    query = null;
                    expectedResults = [];
                }
            } else if (query === null) {
                query = trimmedLine;
            } else {
                expectedResults.push(trimmedLine.split(','));
            }
        }
        if (query) {
            this.addTest(query, expectedResults);
        }
    }
}

(async () => {
    if (process.argv.length < 6) {
        console.error("Usage: node pg_test.js <ip> <port> <user> <password> <testFile>");
        process.exit(1);
    }

    const tests = new Tests();
    await tests.connect(process.argv[2], parseInt(process.argv[3]), process.argv[4], process.argv[5]);
    tests.readTestsFromFile(process.argv[6]);

    if (!(await tests.runTests())) {
        await tests.disconnect();
        process.exit(1);
    }
    await tests.disconnect();
})();

File: compatibility/pg/python/pg_test.py
import psycopg2

class PGTest:
    class Test:
        def __init__(self, query, expected_results):
            self.query = query
            self.expected_results = expected_results

        def run(self, cursor):
            print(f"Running test: {self.query}")
            cursor.execute(self.query)
            if cursor.description is None:
                print("Returns 0 rows")
                return len(self.expected_results) == 0

            rows = cursor.fetchall()
            if len(rows[0]) != len(self.expected_results[0]):
                print(f"Expected {len(self.expected_results[0])} columns, got {len(rows[0])}")
                return False

            for row, expected_row in zip(rows, self.expected_results):
                if list(map(str, row)) != list(map(str, expected_row)):
                    print(f"Expected: {list(map(str, expected_row))}, got: {list(map(str, row))}")
                    return False

            print(f"Returns {len(rows)} rows")
            if len(rows) != len(self.expected_results):
                print(f"Expected {len(self.expected_results)} rows")
                return False

            return True

    def __init__(self):
        self.conn = None
        self.cursor = None
        self.tests = []

    def connect(self, ip, port, user, password):
        try:
            self.conn = psycopg2.connect(
                host=ip,
                port=port,
                dbname="postgres",
                user=user,
                password=password
            )
            self.cursor = self.conn.cursor()
        except Exception as e:
            raise RuntimeError(e)

    def disconnect(self):
        self.cursor.close()
        self.conn.close()

    def add_test(self, query, expected_results):
        self.tests.append(self.Test(query, expected_results))

    def run_tests(self):
        for test in self.tests:
            try:
                self.conn.autocommit = False
                if not test.run(self.cursor):
                    self.conn.rollback()
                    return False
                self.conn.commit()
            except Exception as e:
                print(f"Error running test: {e}")
                self.conn.rollback()
                return False
        return True

    def read_tests_from_file(self, filename):
        with open(filename, "r") as file:
            lines = file.readlines()

        query = None
        expected_results = []
        for line in lines:
            line = line.strip()
            if not line:
                if query:
                    self.add_test(query, expected_results)
                    query = None
                    expected_results = []
            elif query is None:
                query = line
            else:
                expected_results.append(line.split(","))

        if query:
            self.add_test(query, expected_results)

if __name__ == "__main__":
    import sys
    if len(sys.argv) != 6:
        print(f"Usage: {sys.argv[0]} <ip> <port> <user> <password> <testFile>")
        sys.exit(1)

    pg_test = PGTest()
    pg_test.connect(sys.argv[1], int(sys.argv[2]), sys.argv[3], sys.argv[4])
    pg_test.read_tests_from_file(sys.argv[5])

    if not pg_test.run_tests():
        pg_test.disconnect()
        sys.exit(1)

    pg_test.disconnect()